akka-actor/src/main/java/akka/dispatch/AbstractNodeQueue.java

```diff
@@ -47,12 +47,12 @@ protected AbstractNodeQueue() {
     protected final Node<T> peekNode() {
         final Node<T> tail = ((Node<T>)Unsafe.instance.getObjectVolatile(this, tailOffset));
         Node<T> next = tail.next();
-        if (next == null && get() != tail) {
```

Comment:
next is only null if the queue is empty, but then head (`get`) points to the same `Node` as tail, or there is an add in progress which is when we do this hot spin. As soon as the thread (or one of the threads) currently writing to the queue completes setting next we'll progress here

Reply:
and old code will loop and consume CPU until someone wrote something here.

I guess it is just waste of CPU, don't you?

Reply:
No, it will only loop if a write is in progress concurrently with the read. The logic trades a short period of hot spin for short latency and being lock free.

Reply:
@johanandren at my case it loops forever :(